src/svg/qsvggenerator.cpp

```diff
@@ -997,8 +1007,37 @@ void QSvgPaintEngine::updateState(const QPaintEngineState &state)
     // close old state and start a new one...
     if (d->afterFirstUpdate)
         *d->stream << "</g>\n\n";
+    
+    QPainter* p = painter();
+    if (p->hasClipping()) {
+        QPainterPath painter_path = p->clipPathF();
+
+        if (painter_path.elementCount() > 0) {
+            QString clip_path = createPath(painter_path);
+
+            bool path_does_not_exist = 0 == clip_path_to_id.count(clip_path);
+            if (path_does_not_exist) {
+                clip_path_to_id[clip_path] = clip_counter++;
+                
+                d->stream->setString(&d->defs);
```

Comment:
Having looked again at other places in this file, I think you should copy their approach to streaming to `defs` and do:

`QTextStream str(&d_func()->defs, QIODevice::Append);`

and stream to `str`. That way you don't have to reset the stream to body again afterwards. You can't 100% guarantee it was set to body before (though it seems pretty likely).

I'd also recommend moving everything between line 1013 `QPainterPath painter_path = p->clipPathF();` and line 1029 (i.e. everything under 'if hasClipping' except for the last line that streams the `<g` start tag) to a separate helper function similar to the save helpers like `void saveLinearGradientBrush(const QGradient *g)` etc.

In this case `void saveClipPath(const QPainterPath *path)` does the work of converting the path to string, checking if it already exists, incrementing the clip ID and adding it to the lookup, and streaming the painter path to defs. That leaves `updateState` to just check if clipping exists, and if so set the right attribute on the `<g>` tag.